sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+constant representing `A1 && A2`.  The semantic of this decoration is that the
+JIT compiler must discard the function unless the value of the specialization
+constant is `True`.  Augmenting the example from above:
+
+```
+OpDecorate %11 SpecId 1                       ; External ID for spec const A1
+OpDecorate %12 SpecId 2                       ; External ID for spec const A2
+OpDecorate %16 ConditionalINTEL %13           ; Says to discard the function
+                                              ; below when (A1 && A2) is False
+%10 = OpTypeBool
+%11 = OpSpecConstantTrue %10                  ; Represents A1
+%12 = OpSpecConstantTrue %10                  ; Represents A2
+%13 = OpSpecConstantOp %10 LogicalAnd %11 %12 ; Represents A1 && A2
+%14 = OpTypeVoid
+%15 = OpTypeFunction %14
+
+%16 = OpFunction %14 None %15                 ; Definition of function that is
+...                                           ; discarded when (A1 && A2) is False
+OpFunctionEnd
+```
+
+See the extension specification of [SpecConditional][6] for a full
+description of this new SPIR-V decoration.
+
+[6]: <extensions/SPIRV/SPV_INTEL_spec_conditional.asciidoc>
+
+#### Representation in LLVM IR
+
+**TODO**: I need some help here on how to represent the `[[sycl::requires()]]`
+attributes in LLVM IR.  I suspect there is already some mechanism for
+representing SYCL attributes in LLVM IR, so hopefully we can mostly reuse that
+mechanism.
+
+#### Modifications to the post-link tool
+
+The post-link tool must be modified to add the SPIR-V `ConditionalINTEL`
+decorations to the appropriate functions and to emit the specialization
+constants that these decorations reference.  This can be done with two passes
+over the static call tree for each kernel and each exported device function.
+
+**NOTE**: In this context, "exported device function" means a device function
+that is exported from a shared library as defined by [Device Code Dynamic
+Linking][7].
+
+[7]: <https://github.com/intel/llvm/pull/3210>
+
+The first pass operates on each kernel and each exported device function,
+iterating over all the functions in the static call tree of that kernel or
+exported device function to form the union of all aspects it requires.  If the
+kernel or exported device function is decorated with the
+`[[sycl::requires()]]`, those aspects are also added to the union.
+
+**NOTE**: This first pass traverses the static call tree even for kernels that
+are decorated with the `[[sycl::requires()]]` attribute.  Although the
+front-end compiler has already verified that the kernel doesn't require any
+aspects beyond those listed in the attribute, the front-end compiler was only
+able to verify this for the device functions that reside in the same
+translation unit as the kernel.  Therefore, we might still find more required
+aspects in the post-link tool because we have visibility to all device
+functions in the kernel's static call tree, even those that are defined in
+other translation units.
```

Comment:
Yes, this is a good point.  I think we can emit an error instead.

Reply:
To optimize the implementation a bit, we could complete avoid those extra call graph traversals and just say that such situation is a UB

Reply:
We need to write the code to do the graph traversal anyways because it is needed for the kernels that do not have the `[[sycl::requires()]]` attribute.  I expect that most of the kernels will not have that attribute, so skipping the graph traversal just for the ones that have the attribute will not save that much time.  It seems like it would be worth the extra traversal time in these cases in order to get the better error checking.

Reply:
I added wording in 2b51cf572fb0b92f7257968dcdad589639ab7bfb to say that the post-link tool raises a diagnostic if the cross-TU static call tree uses any aspects that are not listed in the `[[sycl::requires()]]` attribute.